include/bbp/sonata/compartment_sets.h

```diff
@@ -0,0 +1,169 @@
+#pragma once
+
+#include <bbp/sonata/nodes.h>
+
+namespace bbp {
+namespace sonata {
+namespace detail {
+class CompartmentLocation;
+class CompartmentSetFilteredIterator;
+class CompartmentSet;
+class CompartmentSets;
+}  // namespace detail
+/**
+ * CompartmentLocation public API.
+ * 
+ * This class uniquely identifies a compartment by a set of gid, section_idx and offset:
+ * 
+ * - gid: Global ID of the cell (Neuron) to which the compartment belongs. No 
+ * overlaps among populations.
+ * - section_idx: Absolute section index. Progressive index that uniquely identifies the section.
+ *  There is a mapping between neuron section names (i.e. dend[10]) and this index.
+ * - offset: Offset of the compartment along the section. The offset is a value between 0 and 1
+ */
+class SONATA_API CompartmentLocation
+{
+  public:
+    CompartmentLocation();
+    CompartmentLocation(const int64_t gid, const int64_t section_idx, const double offset);
+    explicit CompartmentLocation(const std::string& content);
+    explicit CompartmentLocation(std::unique_ptr<detail::CompartmentLocation>&& impl);
+    CompartmentLocation(const CompartmentLocation& other);
+    CompartmentLocation& operator=(const CompartmentLocation& other);
+    CompartmentLocation(CompartmentLocation&&) noexcept;
+    CompartmentLocation& operator=(CompartmentLocation&&) noexcept;
+    ~CompartmentLocation();
+
+    bool operator==(const CompartmentLocation& other) const noexcept;
+    bool operator!=(const CompartmentLocation& other) const noexcept;
+
+    uint64_t gid() const;
+    uint64_t sectionIdx() const;
+    double offset() const;
+
+    std::string toJSON() const;
+
+  private:
+    std::unique_ptr<detail::CompartmentLocation> impl_;
+};
+
+class SONATA_API CompartmentSetFilteredIterator {
+public:
+    using iterator_category = std::input_iterator_tag;
+    using value_type = CompartmentLocation;
+    using difference_type = std::ptrdiff_t;
+    using pointer = void;                // dereference returns by value
+    using reference = CompartmentLocation; // dereference returns by value
+
+    explicit CompartmentSetFilteredIterator(std::unique_ptr<detail::CompartmentSetFilteredIterator> impl);
+    CompartmentSetFilteredIterator(const CompartmentSetFilteredIterator& other);
+    CompartmentSetFilteredIterator& operator=(const CompartmentSetFilteredIterator& other);
+    CompartmentSetFilteredIterator(CompartmentSetFilteredIterator&&) noexcept;
+    CompartmentSetFilteredIterator& operator=(CompartmentSetFilteredIterator&&) noexcept;
+    ~CompartmentSetFilteredIterator();
+
+    /// Dereference operator. It makes a copy!
+    CompartmentLocation operator*() const;
```

Comment:
I think it could return refs; I believe the general expectation in cpp is if you have an iterator, you must keep the container around (and not change it, but we're lucky we're in read only land here), so it seems safe to me.

Reply:
it cannot return a reference. CompartmentLocation has a unique_ptr to the implementation and the original value is in a vector<detail::CompartmentLocation>. I cannot create a unique_ptr on that memory location directly because the owner is the vector. 

If you mean a refrence to the CompartmentLocation that would be a reference to a rvalue which may create undefined behavior. 

Even if that was possible CompartmentLocation has just a unique_ptr and passing by value may be even faster or euql to passing by reference.

Reply:
> CompartmentLocation has a unique_ptr to the implementation

Ah, right, we're running into the the pimpl problem.

I think some of this could be simplified if `CompartmentLocation` was a public `Plain Old Data` struct, what do you think?